anonymous_helpline_chatbot/core/invitations.py

```diff
@@ -1,89 +1,131 @@
+        if sent_message_id is not None:
+            cursor.execute("INSERT INTO sent_invitations(operator_chat_id, client_chat_id, invitation_message_id) "
+                           "VALUES (%s, %s, %s)",
+                           (operator_chat_id, client_chat_id, sent_message_id))
+
+    def invite_to_client(self, client_chat_id: int) -> None:
         """
         Sends out invitation messages to all currently free operators, via which they can start a conversation with the
         client
 
-        :param tg_client_id: Telegram identifier of the user to invite operators to chat with
-        :return: Error code, either of `0`, `1`, `2`, `3`, where `0` indicates that the invitations have been sent
-            successfully, `1` tells that the user had requested invitations before, `2` indicates that there are no free
-            operators, `3` means that the client is in a conversation already (either as a client or as an operator)
+        :param client_chat_id: Messenger identifier of the user to invite operators to have conversation with
         """
-        local_client_id = self.users_controller.get_local_id(tg_client_id)
-
-        with self.conversations_controller.conversations_starter_finisher_lock:
-            free_operators = set(self.users_controller.get_free_operators()) - {tg_client_id}
-            if not free_operators:
-                return 2
-
-            if tg_client_id in self._operators_invitations_messages.keys():
-                return 1
-
-            # In a conversation already
-            if self.conversations_controller.get_conversing(tg_client_id) != ((None, None), (None, None)):
-                return 3
-
-            msg_ids = []
-            for tg_operator_id in free_operators:
-                try:
-                    msg_ids.append((
-                        tg_operator_id,
-                        self.send_invitation_callback(tg_operator_id, tg_client_id,
-                                                      f"Пользователь №{local_client_id} хочет побеседовать. Нажмите "
-                                                      "кнопку ниже, чтобы стать его оператором")
-                    ))
-                except telebot.apihelper.ApiException:
-                    print("Telegram API Exception while sending out operators invitations:", file=stderr)
-                    print(format_exc(), file=stderr)
-
-            self._operators_invitations_messages[tg_client_id] = msg_ids
-
-        return 0
-
-    def clear_invitation_messages(self, tg_client_id: int) -> bool:
+        with self._conn_pool.PrettyCursor() as cursor:
+            # Prevent any invitations from being sent or deleted by parallel transactions until this one completes,
+            # because, for example, a parallel transaction might try to delete an invitation which we are going to send,
+            # but have not sent yet
```

Comment:
This is an ok explanation already, but I think the following one would be even better:
```suggestion
            # Prevent any invitations from being sent or deleted by parallel transactions until this one completes,
            # because, for example, a parallel transaction might try to clear invitations for a user, for which we **are
            # about** to send an invitation to, but have not sent yet
```